In [ ]:
!pip install gensim

In [ ]:
from gensim.corpora import Dictionary
from gensim.models import LsiModel, Phrases, LdaModel, TfidfModel
from gensim.utils import simple_preprocess
from gensim.models import CoherenceModel
import os
import pandas as pd
import numpy as np
import re

In [ ]:
df_clean = pd.read_csv("/content/drive/MyDrive/FYP/Data/reddit-nonislamic/clean.csv")

In [ ]:
from gensim.models import CoherenceModel

tokenized_data = df_clean['document'].apply(lambda x: simple_preprocess(x))

dictionary = Dictionary(tokenized_data)
print("Before removing extremes", len(dictionary))
dictionary.filter_extremes(no_below = 5, no_above = 0.50)
print("After removing extremes", len(dictionary))

corpus = [dictionary.doc2bow(doc) for doc in tokenized_data]

Before removing extremes 93709
After removing extremes 23927


In [ ]:
ldamodel = LdaModel(corpus=corpus, id2word=dictionary, num_topics=45, chunksize=2000, passes=10, iterations=20, alpha='auto', eta="auto", random_state=0)

In [ ]:
coherence_model = CoherenceModel(model=ldamodel, corpus=corpus, dictionary=dictionary, coherence='u_mass')
print("Umass:", coherence_model.get_coherence())

coherence_model = CoherenceModel(model=ldamodel, dictionary=dictionary, texts=tokenized_data, coherence='c_v')
print("Coherence:", coherence_model.get_coherence(), "\n")

Umass: -3.1884408403550846
Coherence: 0.6245461273732219 



In [ ]:
from gensim.test.utils import datapath
import os

folder_name = "lda(nonislamic)-45topics-standardfilter-10pass-20iters-0.625cv"
os.makedirs(f"/content/drive/MyDrive/FYP/Models/{folder_name}", exist_ok=True)

temp_file = datapath(f"/content/drive/MyDrive/FYP/Models/{folder_name}/model")
ldamodel.save(temp_file)

# Additional Stuff

In [ ]:
from pyLDAvis import gensim

pyLDAvis.enable_notebook()
pyLDAvis.gensim.prepare(ldamodel, corpus, dictionary)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
7     -0.247149 -0.011310       1        1  18.196192
45    -0.196751  0.001619       2        1   4.563567
25    -0.178363 -0.168138       3        1   4.310132
13    -0.169753  0.025655       4        1   3.760202
28    -0.202427  0.145008       5        1   3.618207
30    -0.101797 -0.078767       6        1   3.607005
23    -0.109270 -0.246235       7        1   3.318230
40    -0.130837 -0.120918       8        1   3.173016
38    -0.101890  0.180827       9        1   3.094480
41    -0.138095 -0.128707      10        1   3.017890
39    -0.109492 -0.219688      11        1   2.970605
18    -0.170454  0.147782      12        1   2.914752
6      0.061571  0.024931      13        1   2.593638
11     0.005392  0.086977      14        1   2.531063
0     -0.078514 -0.032930      15        1   2.362451
5     -0.056879  0.006609      16        1   2.059749
29    -0.033144  0.114374      17        1   2.039578
36    -0.077660 -0.009350      18        1   1.862621
48    -0.029813 -0.124849      19        1   1.720314
31    -0.134359  0.065023      20        1   1.707693
46    -0.076059  0.095333      21        1   1.690839
32    -0.077940  0.144680      22        1   1.575742
43    -0.026047  0.227854      23        1   1.571743
16    -0.097006  0.009904      24        1   1.567710
1     -0.043118 -0.076938      25        1   1.445450
27    -0.016407  0.018440      26        1   1.440093
19     0.041925 -0.133106      27        1   1.323014
33    -0.009109 -0.133056      28        1   1.235719
47    -0.016298 -0.050749      29        1   1.136505
14     0.067343 -0.025160      30        1   1.110221
17     0.007702 -0.009867      31        1   1.034437
24    -0.013916  0.149194      32        1   1.032327
34    -0.036349  0.065783      33        1   1.020897
3      0.095679  0.073837      34        1   0.950183
22     0.016290  0.066286      35        1   0.777167
12     0.124115  0.023460      36        1   0.732339
4      0.169355 -0.021216      37        1   0.703372
15     0.085949  0.065796      38        1   0.676267
20     0.080216 -0.010123      39        1   0.623246
26     0.183997  0.019609      40        1   0.622268
2      0.130098 -0.039825      41        1   0.566055
49     0.197077 -0.008143      42        1   0.561299
8      0.181209 -0.029524      43        1   0.498180
42     0.174498 -0.031010      44        1   0.426028
35     0.141020  0.021364      45        1   0.424097
10     0.182233 -0.045032      46        1   0.413118
9      0.160491 -0.006947      47        1   0.378463
44     0.184860  0.006524      48        1   0.377059
21     0.190369 -0.009220      49        1   0.367502
37     0.197507 -0.016068      50        1   0.297276, topic_info=             Term          Freq         Total Category  logprob  loglift
21           live  29750.000000  29750.000000  Default  30.0000  30.0000
274          game  22072.000000  22072.000000  Default  29.0000  29.0000
611           day  21774.000000  21774.000000  Default  28.0000  28.0000
256         watch  16267.000000  16267.000000  Default  27.0000  27.0000
71           like  40155.000000  40155.000000  Default  26.0000  26.0000
...           ...           ...           ...      ...      ...      ...
6083       rescue    174.374138    210.904889  Topic50  -4.2747   5.6281
7357          fri    112.654645    126.187511  Topic50  -4.7115   5.7048
3043     duration    143.680173    195.631136  Topic50  -4.4683   5.5096
1265       redeem    119.002161    145.468110  Topic50  -4.6567   5.6174
2142  inspiration    123.799479    179.470370  Topic50  -4.6172   5.4469

[3117 rows x 6 columns], token_table=      Topic      Freq    Term
term                         
5511     37  0.996226     aaa
408       1  0.392406    able
408       2  0.003203    able
408       3  0.066202    able
408       4  0.255019    able
...     ...       ...    

    ldamodel = LdaModel(corpus=corpus, id2word=dictionary, num_topics=i, chunksize=2000, passes=10, iterations=20, alpha='auto', eta="auto", random_state=0)
* also applied default filter extremes


Umass: -2.9096760190107944
Coherence: 0.6150299959812153

Umass: -3.1884408403550846
Coherence: 0.6245461273732219

Umass: -3.633467400923032
Coherence: 0.5792263265336197

Umass: -3.7294564585431416
Coherence: 0.5850464786993977

Umass: -4.079894631330114
Coherence: 0.5783543558070989

Umass: -3.9653513345174516
Coherence: 0.5781129177270508

Umass: -4.700309004627394
Coherence: 0.563148359522355

Umass: -4.742462733754567
Coherence: 0.5430904286425543

Umass: -5.371742339056405
Coherence: 0.5379966749192622

Umass: -4.478230984192431
Coherence: 0.5825752007742571

Umass: -5.213134843034268
Coherence: 0.548329268685571

Umass: -5.380181708151461
Coherence: 0.5357694307050919

Umass: -5.5363572059261905
Coherence: 0.512578110654378